File: app/src/main/java/knf/kuma/player/CustomExoPlayer.java
package knf.kuma.player;

import android.annotation.TargetApi;
import android.app.PictureInPictureParams;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Build;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.util.Rational;
import android.view.View;
import android.widget.ImageButton;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.crashlytics.android.Crashlytics;
import com.google.android.exoplayer2.C;
import com.google.android.exoplayer2.ExoPlaybackException;
import com.google.android.exoplayer2.ExoPlayerFactory;
import com.google.android.exoplayer2.PlaybackParameters;
import com.google.android.exoplayer2.Player;
import com.google.android.exoplayer2.SimpleExoPlayer;
import com.google.android.exoplayer2.Timeline;
import com.google.android.exoplayer2.source.ConcatenatingMediaSource;
import com.google.android.exoplayer2.source.ExtractorMediaSource;
import com.google.android.exoplayer2.source.MediaSource;
import com.google.android.exoplayer2.source.TrackGroupArray;
import com.google.android.exoplayer2.trackselection.AdaptiveTrackSelection;
import com.google.android.exoplayer2.trackselection.DefaultTrackSelector;
import com.google.android.exoplayer2.trackselection.TrackSelection;
import com.google.android.exoplayer2.trackselection.TrackSelectionArray;
import com.google.android.exoplayer2.trackselection.TrackSelector;
import com.google.android.exoplayer2.ui.AspectRatioFrameLayout;
import com.google.android.exoplayer2.ui.SimpleExoPlayerView;
import com.google.android.exoplayer2.upstream.BandwidthMeter;
import com.google.android.exoplayer2.upstream.DefaultBandwidthMeter;
import com.google.android.exoplayer2.upstream.DefaultHttpDataSourceFactory;
import com.google.android.exoplayer2.upstream.FileDataSourceFactory;
import com.google.android.exoplayer2.util.Util;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import knf.kuma.R;
import knf.kuma.commons.EAHelper;
import knf.kuma.database.CacheDB;
import knf.kuma.pojos.QueueObject;
import xdroid.toaster.Toaster;

public class CustomExoPlayer extends AppCompatActivity implements Player.EventListener {
    @BindView(R.id.player)
    SimpleExoPlayerView exoPlayerView;
    @BindView(R.id.title)
    TextView title;
    @BindView(R.id.pip)
    ImageButton pip;
    @BindView(R.id.exo_next)
    ImageButton next;
    @BindView(R.id.exo_prev)
    ImageButton prev;
    @BindView(R.id.progress)
    ProgressBar progressBar;

    @BindView(R.id.lay_top)
    View top;
    @BindView(R.id.lay_bottom)
    View bottom;
    @BindView(R.id.pip_exit)
    ImageButton pip_exit;

    SimpleExoPlayer player;
    private long currentPosition = C.TIME_UNSET;
    private int resumeWindow = C.INDEX_UNSET;

    private int listPosition = 0;
    private List<QueueObject> playList = new ArrayList<>();

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        setTheme(EAHelper.getTheme(this));
        super.onCreate(savedInstanceState);
        setContentView(R.layout.exo_player);
        ButterKnife.bind(this);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N && getPackageManager().hasSystemFeature(PackageManager.FEATURE_PICTURE_IN_PICTURE))
            pip.setVisibility(View.VISIBLE);
        hideUI();
        exoPlayerView.setResizeMode(getResizeMode());
        exoPlayerView.requestFocus();
        if (savedInstanceState != null)
            currentPosition = savedInstanceState.getLong("position", C.TIME_UNSET);
        checkPlaylist(getIntent());
        initPlayer(getIntent());
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putLong("position", currentPosition);
    }

    private void hideUI() {
        getWindow().getDecorView().setSystemUiVisibility(
                View.SYSTEM_UI_FLAG_LAYOUT_STABLE
                        | View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION
                        | View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
                        | View.SYSTEM_UI_FLAG_HIDE_NAVIGATION
                        | View.SYSTEM_UI_FLAG_FULLSCREEN
                        | View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY);
    }

    private int getResizeMode() {
        switch (Objects.requireNonNull(PreferenceManager.getDefaultSharedPreferences(this).getString("player_resize", "0"))) {
            default:
            case "0":
                return AspectRatioFrameLayout.RESIZE_MODE_FIT;
            case "1":
                return AspectRatioFrameLayout.RESIZE_MODE_FILL;
            case "2":
                return AspectRatioFrameLayout.RESIZE_MODE_FIXED_WIDTH;
            case "3":
                return AspectRatioFrameLayout.RESIZE_MODE_FIXED_HEIGHT;
        }
    }

    private void initPlayer(Intent intent) {
        if (player == null) {
            title.setText(intent.getStringExtra("title"));
            BandwidthMeter bandwidthMeter = new DefaultBandwidthMeter();
            TrackSelection.Factory videoTrackSelectionFactory =
                    new AdaptiveTrackSelection.Factory(bandwidthMeter);
            TrackSelector trackSelector =
                    new DefaultTrackSelector(videoTrackSelectionFactory);
            MediaSource source;
            if (intent.getBooleanExtra("isPlayList", false)) {
                List<MediaSource> sourceList = new ArrayList<>();
                playList = CacheDB.INSTANCE.queueDAO().getAllByAid(intent.getStringExtra("playlist"));
                title.setText(playList.get(0).getTitle());
                for (QueueObject object : playList) {
                    if (object.isFile)
                        sourceList.add(new ExtractorMediaSource.Factory(new FileDataSourceFactory()).createMediaSource(object.uri));
                    else
                        sourceList.add(new ExtractorMediaSource.Factory(new DefaultHttpDataSourceFactory(Util.getUserAgent(this, "UKIKU"))).createMediaSource(object.uri));
                }
                source = new ConcatenatingMediaSource(sourceList.toArray(new MediaSource[]{}));
            } else if (!intent.getBooleanExtra("isFile", false)) {
                source = new ExtractorMediaSource.Factory(new DefaultHttpDataSourceFactory(Util.getUserAgent(this, "UKIKU"))).createMediaSource(intent.getData());
            } else {
                source = new ExtractorMediaSource.Factory(new FileDataSourceFactory()).createMediaSource(intent.getData());
            }
            player = ExoPlayerFactory.newSimpleInstance(this, trackSelector);
            exoPlayerView.setPlayer(player);
            player.addListener(this);

            boolean canResume = currentPosition != C.TIME_UNSET;
            if (canResume)
                player.seekTo(currentPosition);
            player.prepare(source, !canResume, false);
            player.setPlayWhenReady(true);

        }
    }

    private void releasePlayer() {
        if (player != null) {
            player.stop();
            player.release();
            player = null;
        }
    }

    @OnClick(R.id.pip)
    @TargetApi(Build.VERSION_CODES.N)
    void onPip(View view) {
        try {
            if (!isInPictureInPictureMode()) {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                    currentPosition = player.getCurrentPosition();
                    PictureInPictureParams params = new PictureInPictureParams.Builder()
                            .setAspectRatio(new Rational(exoPlayerView.getWidth(), exoPlayerView.getHeight()))
                            .build();
                    enterPictureInPictureMode(params);
                } else {
                    enterPictureInPictureMode();
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @OnClick(R.id.skip)
    void onSkip(View view) {
        if (player != null)
            player.seekTo(player.getCurrentWindowIndex(), player.getCurrentPosition() + 85000);
    }

    @OnClick(R.id.pip_exit)
    void onExitPip(View view) {
        getApplication().startActivity(new Intent(this, getClass())
                .putExtra("isReorder", true)
                .addFlags(Intent.FLAG_ACTIVITY_REORDER_TO_FRONT));
    }

    @Override
    public void onPictureInPictureModeChanged(boolean isInPictureInPictureMode) {
        super.onPictureInPictureModeChanged(isInPictureInPictureMode);
        if (!isInPictureInPictureMode) {
            runOnUiThread(() -> {
                top.setVisibility(View.VISIBLE);
                bottom.setVisibility(View.VISIBLE);
                exoPlayerView.setUseController(true);
            });
            /*getApplication().startActivity(new Intent(this, getClass())
                    .putExtra("isReorder", true)
                    .addFlags(Intent.FLAG_ACTIVITY_REORDER_TO_FRONT));*/
        } else {
            runOnUiThread(() -> {
                top.setVisibility(View.GONE);
                bottom.setVisibility(View.GONE);
                progressBar.setVisibility(View.GONE);
                exoPlayerView.setUseController(false);
            });
        }
    }

    private void checkPlaylist(Intent intent) {
        if (!intent.getBooleanExtra("isPlayList", false)) {
            next.setVisibility(View.GONE);
            prev.setVisibility(View.GONE);
        } else {
            next.setVisibility(View.VISIBLE);
            prev.setVisibility(View.VISIBLE);
        }
    }

    @Override
    protected void onNewIntent(Intent intent) {
        setIntent(intent);
        releasePlayer();
        currentPosition = 0;
        resumeWindow = C.INDEX_UNSET;
        checkPlaylist(intent);
        initPlayer(intent);
        super.onNewIntent(intent);
    }

    @Override
    protected void onStart() {
        initPlayer(getIntent());
        super.onStart();
    }

    @Override
    protected void onResume() {
        initPlayer(getIntent());
        super.onResume();
    }

    @Override
    protected void onPause() {
        super.onPause();
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N &&
                isInPictureInPictureMode())
            return;
        if (player != null) {
            resumeWindow = player.getCurrentWindowIndex();
            currentPosition = player.getCurrentPosition();
            releasePlayer();
        }
    }

    @Override
    protected void onStop() {
        releasePlayer();
        super.onStop();
    }

    @Override
    public void onTimelineChanged(Timeline timeline, Object manifest, int reason) {

    }

    @Override
    public void onTracksChanged(TrackGroupArray trackGroups, TrackSelectionArray trackSelections) {

    }

    @Override
    public void onLoadingChanged(final boolean isLoading) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N && isInPictureInPictureMode())
            return;
        progressBar.post(() -> progressBar.setVisibility(isLoading ? View.VISIBLE : View.GONE));
    }

    @Override
    public void onPlayerStateChanged(boolean playWhenReady, int playbackState) {
        if (playbackState == Player.STATE_ENDED) {
            finish();
        }
    }

    @Override
    public void onRepeatModeChanged(int repeatMode) {

    }

    @Override
    public void onShuffleModeEnabledChanged(boolean shuffleModeEnabled) {

    }

    @Override
    public void onPlayerError(ExoPlaybackException error) {
        Exception exception = null;
        switch (error.type) {
            case ExoPlaybackException.TYPE_RENDERER:
                exception = error.getRendererException();
                break;
            case ExoPlaybackException.TYPE_UNEXPECTED:
                exception = error.getUnexpectedException();
                break;
            case ExoPlaybackException.TYPE_SOURCE:
                exception = error.getSourceException();
                break;
        }
        if (exception != null) {
            Toaster.toast("Error al reproducir: " + exception.getMessage());
            Crashlytics.logException(exception);
        } else {
            Toaster.toast("Error desconocido al reproducir");
        }
        finish();
    }

    @Override
    public void onPositionDiscontinuity(int reason) {
        try {
            int latestPosition = player.getCurrentWindowIndex();
            if (latestPosition != listPosition) {
                listPosition = latestPosition;
                title.setText(playList.get(listPosition).getTitle());
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @Override
    public void onPlaybackParametersChanged(PlaybackParameters playbackParameters) {

    }

    @Override
    public void onSeekProcessed() {

    }
}

File: app/src/main/java/knf/kuma/player/CustomExoPlayer.kt
package knf.kuma.player

import android.annotation.TargetApi
import android.app.PictureInPictureParams
import android.content.Intent
import android.content.pm.ActivityInfo
import android.content.pm.PackageManager
import android.content.res.Configuration
import android.graphics.Color
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.view.View
import android.webkit.MimeTypeMap
import android.widget.TextView
import androidx.annotation.RequiresApi
import androidx.lifecycle.lifecycleScope
import androidx.preference.PreferenceManager
import com.google.android.exoplayer2.*
import com.google.android.exoplayer2.source.MediaSource
import com.google.android.exoplayer2.source.ProgressiveMediaSource
import com.google.android.exoplayer2.source.hls.HlsMediaSource
import com.google.android.exoplayer2.ui.AspectRatioFrameLayout
import com.google.android.exoplayer2.upstream.DefaultHttpDataSource
import com.google.firebase.crashlytics.FirebaseCrashlytics
import knf.kuma.R
import knf.kuma.commons.BypassUtil
import knf.kuma.commons.EAHelper
import knf.kuma.commons.SSLSkipper
import knf.kuma.commons.doOnUI
import knf.kuma.commons.noCrash
import knf.kuma.custom.GenericActivity
import knf.kuma.database.CacheDB
import knf.kuma.databinding.ExoPlayerBinding
import knf.kuma.pojos.QueueObject
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.jetbrains.anko.doAsync
import org.jetbrains.anko.find
import xdroid.toaster.Toaster

class CustomExoPlayer : GenericActivity(), Player.Listener {
    private var exoPlayer: ExoPlayer? = null
    private var playerState: PlayerState = PlayerState()
    private var isEnding = false
    private var playList: List<QueueObject> = ArrayList()
    private val binding by lazy { ExoPlayerBinding.inflate(layoutInflater) }

    private val resizeMode: Int
        get() {
            return when (PreferenceManager.getDefaultSharedPreferences(this).getString("player_resize", "0")) {
                "0" -> AspectRatioFrameLayout.RESIZE_MODE_FIT
                "1" -> AspectRatioFrameLayout.RESIZE_MODE_FILL
                "2" -> AspectRatioFrameLayout.RESIZE_MODE_FIXED_WIDTH
                "3" -> AspectRatioFrameLayout.RESIZE_MODE_FIXED_HEIGHT
                else -> AspectRatioFrameLayout.RESIZE_MODE_FIT
            }
        }

    override fun onCreate(savedInstanceState: Bundle?) {
        setTheme(EAHelper.getTheme())
        super.onCreate(savedInstanceState)
        requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_SENSOR_LANDSCAPE
        setContentView(binding.root)
        window.decorView.setBackgroundColor(Color.BLACK)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N && packageManager.hasSystemFeature(PackageManager.FEATURE_PICTURE_IN_PICTURE))
            find<View>(R.id.pip).visibility = View.VISIBLE
        find<View>(R.id.pip).setOnClickListener { onPip() }
        find<View>(R.id.skip).setOnClickListener { onSkip() }
        hideUI()
        binding.player.resizeMode = resizeMode
        binding.player.requestFocus()
        lifecycleScope.launch(Dispatchers.Main) {
            playerState = withContext(Dispatchers.IO) {
                CacheDB.INSTANCE.playerStateDAO().find(intent.getStringExtra("title") ?: "???")
                    ?: PlayerState()
            }
            if (savedInstanceState != null) {
                playerState.position = savedInstanceState.getLong("position", C.TIME_UNSET)
                playerState.window = savedInstanceState.getInt("listPosition", 0)
            }
            checkPlaylist(intent)
            initPlayer(intent)
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putLong("position", playerState.position)
        if (playerState.window != 0)
            outState.putInt("listPosition", playerState.window)
    }

    private fun hideUI() {
        window.decorView.systemUiVisibility = (View.SYSTEM_UI_FLAG_LAYOUT_STABLE
                or View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION
                or View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
                or View.SYSTEM_UI_FLAG_HIDE_NAVIGATION
                or View.SYSTEM_UI_FLAG_FULLSCREEN
                or View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY)
    }

    private fun initPlayer(intent: Intent) {
        if (exoPlayer == null) {
            SSLSkipper.skip()
            lifecycleScope.launch(Dispatchers.Main) {
                find<TextView>(R.id.video_title).text = intent.getStringExtra("title")
                exoPlayer = ExoPlayer.Builder(this@CustomExoPlayer).build()
                binding.player.player = exoPlayer
                exoPlayer?.addListener(this@CustomExoPlayer)
                addMedia(exoPlayer, intent)
                exoPlayer?.prepare()
                val canResume = playerState.window >= 0 && playerState.position != C.TIME_UNSET
                if (canResume)
                    exoPlayer?.seekTo(playerState.window, playerState.position)
                exoPlayer?.playWhenReady = true
                /*disposable = Observable.interval(1, TimeUnit.SECONDS).map { exoPlayer?.currentPosition?:0 }
                        .subscribeOn(AndroidSchedulers.from(exoPlayer?.applicationLooper, false))
                        .subscribe {
                            if (it > 0)
                                lastPosition = it
                        }*/
            }
        }
    }

    private suspend fun addMedia(player: ExoPlayer?, intent: Intent) {
        player?: return
        if (intent.getBooleanExtra("isPlayList", false)) {
            val sourceList = ArrayList<MediaSource>()
            playList = withContext(Dispatchers.IO) {
                CacheDB.INSTANCE.queueDAO()
                        .getAllByAid(intent.getStringExtra("playlist") ?: "empty")
                }
                noCrash { find<TextView>(R.id.video_title).text = playList[0].title() }
                DefaultHttpDataSource.Factory()
                for (queueObject in playList) {
                    sourceList.add(
                        ProgressiveMediaSource.Factory(
                            DefaultHttpDataSource.Factory().apply {
                                setUserAgent(BypassUtil.userAgent)
                            }
                        ).createMediaSource(MediaItem.fromUri(queueObject.createUri()))
                    )
                }
                player.addMediaSources(sourceList)
            } else {
                if (intent.getBooleanExtra("isFile", false) || !intent.hasExtra("headers")) {
                    player.addMediaItem(MediaItem.fromUri(intent.data ?: Uri.parse("")))
                } else {
                    val httpFactory = DefaultHttpDataSource.Factory().apply {
                        setAllowCrossProtocolRedirects(true)
                        intent.getStringArrayExtra("headers") ?.let { headerArray ->
                            val slices = headerArray.toList().chunked(2)
                            val headers = mutableMapOf<String, String>()
                            slices.forEach {
                                headers[it[0]] = it[1]
                            }
                            setDefaultRequestProperties(headers)
                            if (headers.contains("User-Agent")) {
                                setUserAgent(headers["User-Agent"])
                            } else {
                                setUserAgent(BypassUtil.userAgent)
                            }
                        }?: setUserAgent(BypassUtil.userAgent)
                    }
                    val factory = when(MimeTypeMap.getFileExtensionFromUrl(intent.data?.toString())) {
                        "m3u8" -> HlsMediaSource.Factory(httpFactory)
                        else -> ProgressiveMediaSource.Factory(httpFactory)
                    }
                    player.addMediaSource(
                        factory.createMediaSource(
                            MediaItem.fromUri(
                                intent.data ?: Uri.parse("")
                            )
                        )
                    )
                }
            }
    }

    private fun releasePlayer() {
        exoPlayer?.stop()
        exoPlayer?.release()
        exoPlayer = null
    }

    @TargetApi(Build.VERSION_CODES.N)
    internal fun onPip() {
        try {
            if (!isInPictureInPictureMode) {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                    playerState.position = exoPlayer?.currentPosition ?: 0
                    val params = PictureInPictureParams.Builder()
                            //.setAspectRatio(Rational(player.width, player.height))
                            .build()
                    enterPictureInPictureMode(params)
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }

    }

    private fun onSkip() {
        exoPlayer?.seekTo(
            exoPlayer?.currentWindowIndex ?: 0, (exoPlayer?.currentPosition
                ?: 0) + 85000
        )
    }

    @RequiresApi(Build.VERSION_CODES.O)
    override fun onPictureInPictureModeChanged(
        isInPictureInPictureMode: Boolean,
        newConfig: Configuration
    ) {
        super.onPictureInPictureModeChanged(isInPictureInPictureMode, newConfig)
        if (!isInPictureInPictureMode) {
            runOnUiThread {
                find<View>(R.id.lay_top).visibility = View.VISIBLE
                find<View>(R.id.lay_bottom).visibility = View.VISIBLE
                binding.player.useController = true
            }
            /*getApplication().startActivity(new Intent(this, getClass())
                    .putExtra("isReorder", true)
                    .addFlags(Intent.FLAG_ACTIVITY_REORDER_TO_FRONT));*/
        } else {
            runOnUiThread {
                find<View>(R.id.lay_top).visibility = View.GONE
                find<View>(R.id.lay_bottom).visibility = View.GONE
                find<View>(R.id.progress).visibility = View.GONE
                binding.player.useController = false
            }
        }
    }

    private fun checkPlaylist(intent: Intent) {
        if (!intent.getBooleanExtra("isPlayList", false)) {
            find<View>(com.google.android.exoplayer2.R.id.exo_next).visibility = View.GONE
            find<View>(com.google.android.exoplayer2.R.id.exo_prev).visibility = View.GONE
        } else {
            find<View>(com.google.android.exoplayer2.R.id.exo_next).visibility = View.VISIBLE
            find<View>(com.google.android.exoplayer2.R.id.exo_prev).visibility = View.VISIBLE
        }
    }

    override fun onUserLeaveHint() {
        super.onUserLeaveHint()
        exoPlayer?.playWhenReady = false
    }

    override fun onNewIntent(intent: Intent) {
        setIntent(intent)
        releasePlayer()
        playerState.window = C.INDEX_UNSET
        playerState.position = 0
        checkPlaylist(intent)
        initPlayer(intent)
        super.onNewIntent(intent)
    }

    override fun onResume() {
        doOnUI { hideUI() }
        exoPlayer?.playWhenReady = true
        super.onResume()
    }

    override fun onPause() {
        super.onPause()
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N && isInPictureInPictureMode)
            return
        val state = playerState.apply {
            title = find<TextView>(R.id.video_title).text.toString()
            position = if (!isEnding) {
                exoPlayer?.currentPosition ?: 0
            } else
                0
        }
        doAsync {
            CacheDB.INSTANCE.playerStateDAO().set(state)
        }
        if (!isFinishing)
            exoPlayer?.pause()
        else
            exoPlayer?.stop()
    }

    override fun onDestroy() {
        releasePlayer()
        super.onDestroy()
    }

    override fun onTimelineChanged(timeline: Timeline, reason: Int) {

    }

    override fun onLoadingChanged(isLoading: Boolean) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N && isInPictureInPictureMode)
            return
        find<View>(R.id.progress).post { find<View>(R.id.progress).visibility = if (isLoading) View.VISIBLE else View.GONE }
    }

    override fun onPlayerStateChanged(playWhenReady: Boolean, playbackState: Int) {
        if (playbackState == Player.STATE_READY)
            doOnUI { hideUI() }
        if (playbackState == Player.STATE_ENDED) {
            isEnding = true
            finish()
        }
    }

    override fun onRepeatModeChanged(repeatMode: Int) {

    }

    override fun onShuffleModeEnabledChanged(shuffleModeEnabled: Boolean) {

    }

    override fun onPlayerError(error: PlaybackException) {
        Toaster.toast(
            "Error al reproducir: " + error.message?.replace("%", "%%"),
            emptyArray<Any>()
        )
        /*MaterialDialog(this).show {
            message(text = error.stackTraceToString().also {
                (getSystemService(Context.CLIPBOARD_SERVICE) as? ClipboardManager)?.setPrimaryClip(ClipData.newPlainText("stack", it))
            })
            positiveButton(text = "OK")
        }*/
        FirebaseCrashlytics.getInstance().recordException(error)
        finish()
    }

    /*override fun onPositionDiscontinuity(reason: Int) {
        try {
            val latestPosition = exoPlayer?.currentWindowIndex ?: 0
            if (latestPosition != listPosition) {
                val state = PlayerState().apply {
                    title = playList[listPosition].title()
                    if (reason == 0) {
                        position = 0
                    } else if (reason in 1..2) {
                        position = lastPosition
                    }
                }
                doAsync {
                    CacheDB.INSTANCE.playerStateDAO().set(state)
                }
                listPosition = latestPosition
                video_title.text = playList[listPosition].title()
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }*/

    override fun onPositionDiscontinuity(
        oldPosition: Player.PositionInfo,
        newPosition: Player.PositionInfo,
        reason: Int
    ) {
        try {
            val latestPosition = newPosition.windowIndex
            if (latestPosition != playerState.window) {
                val state = playerState.apply {
                    title = playList[playerState.window].title()
                    if (reason == 0) {
                        position = 0
                    } else if (reason in 1..2) {
                        position = oldPosition.positionMs
                    }
                }
                doAsync {
                    CacheDB.INSTANCE.playerStateDAO().set(state)
                }
                playerState.window = latestPosition
                find<TextView>(R.id.video_title).text = playList[playerState.window].title()
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    override fun onPlaybackParametersChanged(playbackParameters: PlaybackParameters) {

    }
}
